pkg/scheduler/framework/plugins/podtopologyspread/filtering.go

```diff
@@ -280,6 +280,79 @@ func calPreFilterState(pod *v1.Pod, allNodes []*schedulernodeinfo.NodeInfo) (*pr
 	return &s, nil
 }
 
+// fit checks if pod is a good fit on the given node.
+func (pl *PodTopologySpread) fit(pod *v1.Pod, nodeInfo *nodeinfo.NodeInfo) *framework.Status {
+	constraints, err := filterTopologySpreadConstraints(pod.Spec.TopologySpreadConstraints, v1.DoNotSchedule)
+	if err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+
+	podLabelSet := labels.Set(pod.Labels)
+	for _, c := range constraints {
+		// The node won't be a fit if it doesn't have c.topologyKey present.
+		tpKey := c.topologyKey
+		tpVal, ok := nodeInfo.Node().Labels[tpKey]
+		if !ok {
+			klog.V(5).Infof("node '%s' doesn't have required label '%s'", nodeInfo.Node().Name, tpKey)
+			return framework.NewStatus(framework.Unschedulable, ErrReasonConstraintsNotMatch)
+		}
+
+		tpPairToMatchNum := make(map[topologyPair]int32)
+		// Iterate through all nodes.
```

Comment:
Actually should we move 

```
nodeInfos, _ := pl.sharedLister.NodeInfos().List()
		for _, nodeInfo := range nodeInfos {
			// (1) Ensure the node matches the incoming Pod's node selector/affinity, if specified.
			if !pluginhelper.PodMatchesNodeSelectorAndAffinityTerms(pod, nodeInfo.Node()) {
				continue
			}
```

and

```
for _, existingPod := range nodeInfo.Pods() {
				if existingPod.Namespace != pod.Namespace {
					continue
				}
```

out of `for _, c := range constraints` to get a pre-filtered pod list first. Suppose has huge node and several constraints, time complexity would increase a lot due to call `pluginhelper.PodMatchesNodeSelectorAndAffinityTerms(pod, nodeInfo.Node()` every time.

Reply:
Calculating pre-filtered pod list doesn't look right as you can't only focus on matching Pods, instead, you need to also consider the case that nodes can be legit and don't have any matching Pod. In that case, the pre-filtered pod list lost that info.

I **_was_** thinking about calculating a pre-filtered node list, but it's only beneficial when len(constraints) >= 2, and given that it's a slow-path which we don't expect it to be executed by critical paths, so I gave up this idea.

Reply:
got it, thx for detail explaination.